BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,194 @@
+        setContentView(binding.root)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        cryptographyManager = CryptographyManager()
+        val ciphertextWrapper = cryptographyManager.getCiphertextWrapperFromSharedPrefs(
+            applicationContext,
+            SHARED_PREFS_FILENAME,
+            Context.MODE_PRIVATE,
+            CIPHERTEXT_WRAPPER
+        )
+
+        if (ciphertextWrapper != null) {
+            if (SampleAppUser.fakeToken == null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                // The user has already logged in, so proceed to the rest of the app
+                // this is a todo for you, the developer
+                updateApp(getString(R.string.already_signedin))
+            }
+        } else {
+            loginWithPasswordManager()
+        }
+
+        binding.useBiometrics.setOnClickListener {
+            if (ciphertextWrapper != null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                startActivity(Intent(this, EnableBiometricLoginActivity::class.java))
+            }
+        }
+    }
+
+
+    // BIOMETRICS SECTION
+
+    private fun showBiometricPromptForDecryption() {
+        val canAuthenticate = BiometricManager.from(applicationContext).canAuthenticate()
+        if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
+            val secretKeyName = getString(R.string.secret_key_name)
+            val cipher = cryptographyManager.getInitializedCipherForDecryption(
+                secretKeyName, encryptedServerTokenWrapper.initializationVector
+            )
+            biometricPrompt =
+                BiometricPromptUtils.createBiometricPrompt(this, ::decryptServerTokenFromStorage)
+            val promptInfo = BiometricPromptUtils.createPromptInfo(this)
+            biometricPrompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
+        }
+    }
+
+    private fun decryptServerTokenFromStorage(authResult: BiometricPrompt.AuthenticationResult) {
+        authResult?.cryptoObject?.cipher?.apply {
+            val plaintext =
+                cryptographyManager.decryptData(encryptedServerTokenWrapper.ciphertext, this)
+            SampleAppUser.fakeToken = plaintext
+            // now that you have the token, you can query server for everything else
+            // the only reason we call this fakeToken is because we didn't really get it from
+            // the server. In your case, you will have gotten it from the server the first time
+            // and therefore, it's a real token.
+
+            updateApp(getString(R.string.already_signedin))
+        }
+    }
+
+    // USERNAME + PASSWORD SECTION
+
+    private fun loginWithPasswordManager() {
+        val loginWithPasswordViewModel by viewModels<LoginWithPasswordViewModel>()
```

Comment:
This is not the right way to use a ViewModel.

Reply:
Please elaborate

Reply:
Please see the [view model documentation](https://developer.android.com/topic/libraries/architecture/viewmodel) -- view models should be created in `onCreate`.